bibcheck/plugins/journal_names.py

```diff
@@ -106,12 +106,22 @@ def check_variants(record, sort_variants):
 
```

Comment:
remove line 103 `if sort_variants:` and move lines 104-105 below 107
it appears that `sort_index` is used only once on line 109

this appears to be all a convoluted way to sort by name length of journal variant in reverse order
`m730_sort = sorted(m730, key=lambda x:len(x[0][0][1]), reverse=True)`
accomplishes same and doesn't use intermediate dict and doesn't require `sorted_keys` -- which I know I suggested based on initial reading of the code

anyhow, you can leave that as is, just remove the duplicate if statement. 



sorting also has a `reverse=True` option which is more readable than `*-1`

for example
```
$ cat /tmp/namelengthsort.py 
# coding: utf-8
from invenio.bibedit_utils import get_bibrecord
from invenio.bibcheck_task import AmendableRecord
record = AmendableRecord(get_bibrecord(1212848))
x = record['730']
for f in x:
    print(f)

print("\nafter sorting\n")
x = sorted(x, key=lambda x:len(x[0][0][1]), reverse=True)
for f in x:
    print(f)
```

```python
$ python /tmp/namelengthsort.py 
([('a', 'J AUSTRAL MATH SOC SERIESA'), ('b', 'A')], ' ', ' ', '', 10)
([('a', 'Journal of the Australian Mathematical Society. Series A: Pure Mathematics and Statistics J.Austra.Math.Soc.Ser.A'), ('b', 'A')], ' ', ' ', '', 11)
([('a', 'J AUSTRAL MATH SOC')], ' ', ' ', '', 12)
([('a', 'J AUST MATH SOC')], ' ', ' ', '', 13)
([('a', 'J.Austr.Math.Soc.')], ' ', ' ', '', 14)
([('a', 'JAUMA')], ' ', ' ', '', 15)
([('a', 'J.Austral.Math.Soc.Ser.A'), ('b', 'A')], ' ', ' ', '', 16)
([('a', 'J AUSTRAL MATH SOC SER A'), ('b', 'A')], ' ', ' ', '', 17)
([('a', 'J AUST MATH SOC SER B APPL MATH'), ('b', 'B')], ' ', ' ', '', 18)
([('a', 'J AUST MATH SOC B APPL MATH'), ('b', 'B')], ' ', ' ', '', 19)
([('a', 'J AUSTRAL MATH SOC SER B'), ('b', 'B')], ' ', ' ', '', 20)
([('a', 'J AUSTR MATH SOC'), ('b', 'B')], ' ', ' ', '', 21)
([('a', 'JAMAD'), ('b', 'A')], ' ', ' ', '', 22)
([('a', 'JAMMD'), ('b', 'B')], ' ', ' ', '', 23)

after sorting

([('a', 'Journal of the Australian Mathematical Society. Series A: Pure Mathematics and Statistics J.Austra.Math.Soc.Ser.A'), ('b', 'A')], ' ', ' ', '', 11)
([('a', 'J AUST MATH SOC SER B APPL MATH'), ('b', 'B')], ' ', ' ', '', 18)
([('a', 'J AUST MATH SOC B APPL MATH'), ('b', 'B')], ' ', ' ', '', 19)
([('a', 'J AUSTRAL MATH SOC SERIESA'), ('b', 'A')], ' ', ' ', '', 10)
([('a', 'J.Austral.Math.Soc.Ser.A'), ('b', 'A')], ' ', ' ', '', 16)
([('a', 'J AUSTRAL MATH SOC SER A'), ('b', 'A')], ' ', ' ', '', 17)
([('a', 'J AUSTRAL MATH SOC SER B'), ('b', 'B')], ' ', ' ', '', 20)
([('a', 'J AUSTRAL MATH SOC')], ' ', ' ', '', 12)
([('a', 'J.Austr.Math.Soc.')], ' ', ' ', '', 14)
([('a', 'J AUSTR MATH SOC'), ('b', 'B')], ' ', ' ', '', 21)
([('a', 'J AUST MATH SOC')], ' ', ' ', '', 13)
([('a', 'JAUMA')], ' ', ' ', '', 15)
([('a', 'JAMAD'), ('b', 'A')], ' ', ' ', '', 22)
([('a', 'JAMMD'), ('b', 'B')], ' ', ' ', '', 23)
```

mmh, is 730__b:B correct here? see last line

this relies on the subfields being sorted and the subfield tuples being in order `(sf, val)` so that `x[0][0][1]` always is the journal name from `730__a`. I don't know how robust that is in practice.

Reply:
Sorting:
There should be blocks according to $$b. First everything without $$b followed by $$bA, $$bB, ...
Then by length (longest first), alphabetically in case of same length. Only by length would be reversed. https://github.com/inspirehep/inspire/blob/ac457c22c71151dbb34cb3cf5ced495ca59506da/bibcheck/plugins/journal_names.py#L104
I.e.
```
001212848 730__ $$aJ AUSTRAL MATH SOC SERIESA$$bA
001212848 730__ $$aJOURNAL OF THE AUSTRALIAN MATHEMATICAL SOCIETY SERIES A: PURE MATHEMATICS AND STATISTICS J AUSTRA MATH SOC SER A$$bA
001212848 730__ $$aJ AUSTRAL MATH SOC
001212848 730__ $$aJ AUST MATH SOC
001212848 730__ $$aJ AUSTR MATH SOC
001212848 730__ $$aJAUMA
001212848 730__ $$aJ AUSTRAL MATH SOC SER A$$bA
001212848 730__ $$aJ AUSTRAL MATH SOC SER A$$bA
001212848 730__ $$aJ AUST MATH SOC SER B APPL MATH$$bB
001212848 730__ $$aJ AUST MATH SOC B APPL MATH$$bB
001212848 730__ $$aJ AUSTRAL MATH SOC SER B$$bB
001212848 730__ $$aJ AUSTR MATH SOC$$bB
001212848 730__ $$aJAMAD
001212848 730__ $$aJAMMD
(record modified to have an example of alphabetical sorting)
```
->

```
001212848 730__ $$aJ AUSTRAL MATH SOC
001212848 730__ $$aJ AUSTR MATH SOC
001212848 730__ $$aJ AUST MATH SOC
001212848 730__ $$aJAMAD
001212848 730__ $$aJAMMD
001212848 730__ $$aJAUMA
001212848 730__ $$aJOURNAL OF THE AUSTRALIAN MATHEMATICAL SOCIETY SERIES A: PURE MATHEMATICS AND STATISTICS J AUSTRA MATH SOC SER A$$bA
001212848 730__ $$aJ AUSTRAL MATH SOC SERIESA$$bA
001212848 730__ $$aJ AUSTRAL MATH SOC SER A$$bA
001212848 730__ $$aJ AUSTRAL MATH SOC SER A$$bA
001212848 730__ $$aJ AUST MATH SOC SER B APPL MATH$$bB
001212848 730__ $$aJ AUST MATH SOC B APPL MATH$$bB
001212848 730__ $$aJ AUSTRAL MATH SOC SER B$$bB
001212848 730__ $$aJ AUSTR MATH SOC$$bB

Record marked as invalid: variant is listed twice: "J AUSTRAL MATH SOC SER A" 
Record marked as invalid: variant is listed twice: "J AUSTR MATH SOC" 
since there is a $$b subfield this is not done automatically
```

I never understood the `key=lambda...` syntax, I have absolutely no idea what you are doing there. If it relies on order of subfields it will not work.
I'm afraid I can't come up with a solution without `sort_index`.
And to build the index I have to loop over the fields.
Either I have two if-statements or two loops. 

I can replace
https://github.com/inspirehep/inspire/blob/ac457c22c71151dbb34cb3cf5ced495ca59506da/bibcheck/plugins/journal_names.py#L109
by

```
m730_sort = [m730[num_f] for num_f in [k for k, _ in sorted(sort_index.items(), key=itemgetter(1))]]
```
which I find un-readable.
or
```
sorted_keys = [k for k, _ in sorted(sort_index.items(), key=itemgetter(1))]
m730_sort = [m730[num_f] for num_f in sorted_keys]
```


Content:
JAMAD, JAMMD, JAUMA are Codens, we have to check which is which.
`$$aJ AUSTR MATH SOC$$bB` is definitely wrong.